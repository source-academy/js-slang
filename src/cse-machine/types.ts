import * as es from 'estree'

import { Environment } from '../types'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
  srcNode: es.Node
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
