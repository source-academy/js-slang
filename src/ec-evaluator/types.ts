import * as es from 'estree'

import { Context } from '..'
import { Environment } from '../types'
import { Agenda, Stash } from './interpreter'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  PUSH_UNDEFINED_IF_NEEDED = 'PushUndefinedIfNeeded',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
  srcNode: es.Node
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
  srcNode: es.Node
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
  srcNode: es.Node
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
  srcNode: es.Node
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

export type CmdEvaluator = (
  command: AgendaItem,
  context: Context,
  agenda: Agenda,
  stash: Stash
) => void

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
