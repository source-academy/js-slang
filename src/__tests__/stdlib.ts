import { Value } from '../types'
import { expectResult, snapshotFailure, stripIndent } from '../utils/testing'

test.each([
  [
    1,
    `
    display('message');
    `,
    true,
    'message'
  ],

  [
    1,
    `
    error('error!');
    `,
    false,
    undefined
  ],

  [
    1,
    `
    is_undefined(undefined);
    `,
    true,
    true
  ],

  [
    2,
    `
    is_undefined(null);
    `,
    true,
    false
  ],

  [
    2,
    `
    is_null(undefined);
    `,
    true,
    false
  ],

  [
    2,
    `
    is_null(null);
    `,
    true,
    true
  ],

  [
    1,
    `
    is_string('string');
    `,
    true,
    true
  ],

  [
    1,
    `
    is_string('true');
    `,
    true,
    true
  ],

  [
    1,
    `
    is_string('1');
    `,
    true,
    true
  ],

  [
    1,
    `
    is_string(true);
    `,
    true,
    false
  ],

  [
    1,
    `
    is_string(1);
    `,
    true,
    false
  ],

  [
    1,
    `
    is_number('string');
    `,
    true,
    false
  ],

  [
    1,
    `
    is_number('true');
    `,
    true,
    false
  ],

  [
    1,
    `
    is_number('1');
    `,
    true,
    false
  ],

  [
    1,
    `
    is_number(true);
    `,
    true,
    false
  ],

  [
    1,
    `
    is_number(1);
    `,
    true,
    true
  ],

  [
    1,
    `
    is_boolean('string');
    `,
    true,
    false
  ],

  [
    1,
    `
    is_boolean('true');
    `,
    true,
    false
  ],

  [
    1,
    `
    is_boolean('1');
    `,
    true,
    false
  ],

  [
    1,
    `
    is_boolean(true);
    `,
    true,
    true
  ],

  [
    1,
    `
    is_boolean(1);
    `,
    true,
    false
  ],

  [
    1,
    `
    is_function(display);
    `,
    true,
    true
  ],

  [
    1,
    `
    is_function(x => x);
    `,
    true,
    true
  ],

  [
    1,
    `
    function f(x) {
      return x;
    }
    is_function(f);
    `,
    true,
    true
  ],

  [
    1,
    `
    is_function(1);
    `,
    true,
    false
  ],

  [
    3,
    `
    is_array(1);
    `,
    true,
    false
  ],

  [
    3,
    `
    is_array(pair(1, 2));
    `,
    true,
    true
  ],

  [
    3,
    `
    is_array([1]);
    `,
    true,
    true
  ],

  [
    100,
    `
    is_object(1);
    `,
    true,
    false
  ],

  [
    100,
    `
    is_object(pair(1, 2));
    `,
    true,
    true
  ],

  [
    100,
    `
    is_object([1]);
    `,
    true,
    true
  ],

  [
    100,
    `
    is_object({});
    `,
    true,
    true
  ],

  [
    100,
    `
    is_object({a: 1});
    `,
    true,
    true
  ],

  [
    100,
    `
    is_object(x => x);
    `,
    true,
    true
  ],

  [
    100,
    `
    is_object(display);
    `,
    true,
    true
  ],

  [
    100,
    `
    is_object(1);
    `,
    true,
    false
  ],

  [
    100,
    `
    is_object('string');
    `,
    true,
    false
  ],

  [
    100,
    `
    is_object(true);
    `,
    true,
    false
  ],

  [
    100,
    `
    is_NaN(1 / 0);
    `,
    true,
    false
  ],

  [
    100,
    `
    is_NaN(NaN);
    `,
    true,
    true
  ],

  [
    100,
    `
    is_NaN(1);
    `,
    true,
    false
  ],

  [
    100,
    `
    is_NaN(x => x);
    `,
    true,
    false
  ],

  [
    100,
    `
    has_own_property({a: 1, b: 2}, 'a');
    `,
    true,
    true
  ],

  [
    100,
    `
    has_own_property({a: 1, b: 2}, 'c');
    `,
    true,
    false
  ],

  // NOTE: OOP doesn't work because we've disabled inherited properties for security reasons
  // https://github.com/source-academy/js-slang/pull/137
  // https://github.com/source-academy/js-slang/issues/81
  // [
  //   100,
  //   `
  //   let o = {a: 1};
  //   let o2 = {b: 2};
  //   o.__proto__ = o2;
  //   o.b;
  //   `,
  //   true,
  //   2
  // ],
  //
  // [
  //   100,
  //   `
  //   let o = {a: 1};
  //   let o2 = {b: 2};
  //   o.__proto__ = o2;
  //   has_own_property(o, 'a');
  //   `,
  //   true,
  //   true
  // ],
  //
  // [
  //   100,
  //   `
  //   let o = {a: 1};
  //   let o2 = {b: 2};
  //   o.__proto__ = o2;
  //   has_own_property(o, 'b');
  //   `,
  //   true,
  //   false
  // ],

  [
    3,
    `
    array_length([1]);
    `,
    true,
    1
  ],

  [
    1,
    `
    parse_int('10', 10);
    `,
    true,
    10
  ],

  [
    1,
    `
    parse_int('10', 2);
    `,
    true,
    2
  ],

  [
    1,
    `
    is_number(runtime());
    `,
    true,
    true
  ],

  [
    1,
    `
    const start = runtime();
    function repeatUntilDifferentTime() {
      if (start === runtime()) {
        return repeatUntilDifferentTime();
      } else {
        return true;
      }
    }
    repeatUntilDifferentTime();
    `,
    true,
    true
  ],

  [
    2,
    `
    pair(1, 2);
    `,
    true,
    [1, 2]
  ],

  [
    2,
    `
    list(1, 2);
    `,
    true,
    [1, [2, null]]
  ],

  [
    2,
    `
    is_list(1);
    `,
    true,
    false
  ],

  [
    2,
    `
    is_list(pair(1, 2));
    `,
    true,
    false
  ],

  [
    2,
    `
    is_list(list(1, 2));
    `,
    true,
    true
  ],

  [
    2,
    `
    head(pair(1, 2));
    `,
    true,
    1
  ],

  [
    2,
    `
    tail(pair(1, 2));
    `,
    true,
    2
  ],

  [
    2,
    `
    head(null);
    `,
    false,
    undefined
  ],

  [
    2,
    `
    tail(null);
    `,
    false,
    undefined
  ],

  [
    2,
    `
    head(1);
    `,
    false,
    undefined
  ],

  [
    2,
    `
    tail(1);
    `,
    false,
    undefined
  ],

  [
    2,
    `
    length(list(1, 2));
    `,
    true,
    2
  ],

  [
    2,
    `
    length(1);
    `,
    false,
    undefined
  ]
] as Array<[number, string, boolean, Value]>)(
  'Builtins work as expected %#',
  (chapter: number, snippet: string, passing: boolean, returnValue: Value) => {
    if (passing) {
      return expectResult(stripIndent(snippet), { chapter, native: chapter < 100 }).toEqual(
        returnValue
      )
    } else {
      return snapshotFailure(stripIndent(snippet), { chapter }, 'fails')
    }
  }
)
