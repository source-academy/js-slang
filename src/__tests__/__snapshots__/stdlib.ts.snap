// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builtins work as expected 0: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display('message');",
  "displayResult": Array [
    "\\"message\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": "message",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 1: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "error('error!');",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: "error!"],
      "location": SourceLocation {
        "end": Position {
          "column": 15,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: \\"error!\\"",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 2: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_undefined(undefined);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 3: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_undefined(null);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 4: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_null(undefined);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 5: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_null(null);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 6: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string('string');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 7: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string('true');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 8: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string('1');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 9: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(true);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 10: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 11: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number('string');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 12: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number('true');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 13: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number('1');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 14: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(true);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 15: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 16: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean('string');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 17: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean('true');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 18: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean('1');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 19: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(true);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 20: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 21: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(display);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 22: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(x => x);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 23: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function f(x) {
  return x;
}
is_function(f);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 24: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 25: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_array(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 26: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_array(pair(1, 2));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 27: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_array([1]);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 28: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 29: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object(pair(1, 2));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 30: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object([1]);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 31: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object({});",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 32: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object({a: 1});",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 33: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object(x => x);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 34: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object(display);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 35: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 36: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object('string');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 37: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_object(true);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 38: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_NaN(1 / 0);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 39: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_NaN(NaN);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 40: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_NaN(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 41: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_NaN(x => x);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 42: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "has_own_property({a: 1, b: 2}, 'a');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 43: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "has_own_property({a: 1, b: 2}, 'c');",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 44: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "array_length([1]);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 45: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "parse_int('10', 10);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 10,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 46: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "parse_int('10', 2);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 47: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(runtime());",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 48: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const x = runtime();
const f = x => x === 0 ? x : f(x-1);
f(1000);
const y = runtime();
x !== y;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 49: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "pair(1, 2);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": Array [
    1,
    2,
  ],
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 50: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "list(1, 2);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": Array [
    1,
    Array [
      2,
      null,
    ],
  ],
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 51: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_list(1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 52: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_list(pair(1, 2));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 53: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_list(list(1, 2));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 54: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(pair(1, 2));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 55: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "tail(pair(1, 2));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 56: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(null);",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: head(xs) expects a pair as argument xs, but encountered null],
      "location": SourceLocation {
        "end": Position {
          "column": 10,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: head(xs) expects a pair as argument xs, but encountered null",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 57: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "tail(null);",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: tail(xs) expects a pair as argument xs, but encountered null],
      "location": SourceLocation {
        "end": Position {
          "column": 10,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: tail(xs) expects a pair as argument xs, but encountered null",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 58: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(1);",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: head(xs) expects a pair as argument xs, but encountered 1],
      "location": SourceLocation {
        "end": Position {
          "column": 7,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: head(xs) expects a pair as argument xs, but encountered 1",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 59: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "tail(1);",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: tail(xs) expects a pair as argument xs, but encountered 1],
      "location": SourceLocation {
        "end": Position {
          "column": 7,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: tail(xs) expects a pair as argument xs, but encountered 1",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 60: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "length(list(1, 2));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 61: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "length(1);",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: tail(xs) expects a pair as argument xs, but encountered 1],
      "location": SourceLocation {
        "end": Position {
          "column": 9,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: tail(xs) expects a pair as argument xs, but encountered 1",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;
